src/main/java/edu/kit/hci/soli/controller/LoginControllerAdvice.java

```diff
@@ -28,31 +29,35 @@ public CsrfToken getCsrfToken(HttpServletRequest request) {
     }
 
     @ModelAttribute("login")
-    public LoginStateModel getLoginStateModel(Principal principal, @AuthenticationPrincipal OidcUser oidcUser, HttpServletRequest request) {
-        CsrfToken csrfToken = (CsrfToken) request.getAttribute(CsrfToken.class.getName());
-        // TODO what if null?
-
-        //TODO the user should NOT be null, since we need it in templates
+    public LoginStateModel getLoginStateModel(Principal principal, @AuthenticationPrincipal OidcUser oidcUser, HttpServletRequest request, @ModelAttribute("csrf") CsrfToken csrf) {
         if (principal == null) {
-            return new LoginStateModel("Visitor", LoginStateModel.Kind.VISITOR, csrfToken, null);
+            return new LoginStateModel("Visitor", LoginStateModel.Kind.VISITOR, csrf, null);
         }
         if (oidcUser == null) {
-            return new LoginStateModel(principal.getName(), LoginStateModel.Kind.ADMIN, csrfToken, null);
+            User user = userService.resolveAdminUser();
+            if (user == null) {
+                log.error("No admin user found in database, creating new");
+                user = new NonOidcUser();
+                user.setEmail(null);
+                user.setUsername("admin");
+                user.setUserId("admin");
+                userService.create(user);
+            }
+            return new LoginStateModel(principal.getName(), LoginStateModel.Kind.ADMIN, csrf, user);
         }
 
-
         String username = oidcUser.getUserInfo().getFullName();
 
-        String email = oidcUser.getUserInfo().getEmail();
-        User user = userService.findByEmail(email);
+        String id = "kit/" + principal.getName();
+        User user = userService.findByUserId(id);
         if (user == null) {
-            log.info("Creating new OIDC user: {} with email: {}", username, email);
-            user = new User();
-            user.setEmail(email);
+            log.info("Creating new OIDC user: {} with id: {}", username, id);
+            user = new edu.kit.hci.soli.domain.OidcUser();
```

Comment:
Please import it. Is this for some reason intentional?

Reply:
We already import a class called `OidcUser` (used in the method declaration)